flash/core/finetuning.py

```diff
@@ -0,0 +1,151 @@
+            attr = getattr(pl_module, attr_name, None)
+            if attr is None or not isinstance(attr, nn.Module):
+                MisconfigurationException("To use NeverUnfreeze your model must have a {attr} attribute")
+            self.freeze(module=attr, train_bn=self.train_bn)
+
+
+class FreezeUnfreeze(NeverUnfreeze):
+
+    def __init__(
+        self, attr_names: Union[str, List[str]] = "backbone", train_bn: bool = True, unfreeze_at_epoch: int = 10
+    ):
+        super().__init__(attr_names, train_bn)
+        self.unfreeze_at_epoch = unfreeze_at_epoch
+
+    def finetunning_function(
+        self,
+        pl_module: pl.LightningModule,
+        epoch: int,
+        optimizer: Optimizer,
+        opt_idx: int,
+    ) -> None:
+        if epoch == self.unfreeze_at_epoch:
+            modules = []
+            for attr_name in self.attr_names:
+                modules.append(getattr(pl_module, attr_name))
+
+            self.unfreeze_and_add_param_group(
+                module=modules,
+                optimizer=optimizer,
+                train_bn=self.train_bn,
+            )
+
+
+class MilestonesFinetuning(BaseFinetuning):
+
+    def __init__(self, unfreeze_milestones: tuple = (5, 10), train_bn: bool = True, num_layers: int = 5):
+        self.unfreeze_milestones = unfreeze_milestones
+        self.train_bn = train_bn
+        self.num_layers = num_layers
+
+    def freeze_before_training(self, pl_module: pl.LightningModule) -> None:
+        # TODO: might need some config to say which attribute is model
+        # maybe something like:
+        # self.freeze(module=pl_module.getattr(self.feature_attr), train_bn=self.train_bn)
+        # where self.feature_attr can be "backbone" or "feature_extractor", etc.
+        # (configured in init)
+        assert hasattr(pl_module, "backbone"), "To use MilestonesFinetuning your model must have a backbone attribute"
+        self.freeze(module=pl_module.backbone, train_bn=self.train_bn)
+
+    def finetunning_function(
+        self,
+        pl_module: pl.LightningModule,
+        epoch: int,
+        optimizer: Optimizer,
+        opt_idx: int,
+    ) -> None:
+        backbone_modules = list(pl_module.backbone.modules())
+        if epoch == self.unfreeze_milestones[0]:
+            # unfreeze 5 last layers
+            # TODO last N layers should be parameter
+            self.unfreeze_and_add_param_group(
+                module=backbone_modules[-self.num_layers:],
+                optimizer=optimizer,
+                train_bn=self.train_bn,
+            )
+
+        elif epoch == self.unfreeze_milestones[1]:
+            # unfreeze remaining layers
+            # TODO last N layers should be parameter
+            self.unfreeze_and_add_param_group(
+                module=backbone_modules[:-self.num_layers],
+                optimizer=optimizer,
+                train_bn=self.train_bn,
+            )
+
+
+def instantiate_cls(cls, kwargs):
```

Comment:
why do we need this instead of just passing `**kwargs` to the callback?

Reply:
removed !